<a href="https://colab.research.google.com/github/MarkLee7925/Bank-Marketing-Classification/blob/main/BankMarketing_XGB.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Bank Marketing - XGBoost

## Data preprocessing

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
# Load dataframe (df)
df = pd.read_csv('bank-additional-full.csv', sep=';')
X, y = df.iloc[:, :-1].values, df.iloc[:, -1].values

print(X.shape, y.shape)

(41188, 20) (41188,)


In [3]:
df

age          job  marital  ... euribor3m nr.employed    y
0       56    housemaid  married  ...     4.857      5191.0   no
1       57     services  married  ...     4.857      5191.0   no
2       37     services  married  ...     4.857      5191.0   no
3       40       admin.  married  ...     4.857      5191.0   no
4       56     services  married  ...     4.857      5191.0   no
...    ...          ...      ...  ...       ...         ...  ...
41183   73      retired  married  ...     1.028      4963.6  yes
41184   46  blue-collar  married  ...     1.028      4963.6   no
41185   56      retired  married  ...     1.028      4963.6   no
41186   44   technician  married  ...     1.028      4963.6  yes
41187   74      retired  married  ...     1.028      4963.6   no

[41188 rows x 21 columns]

In [4]:
df.dtypes

age                 int64
job                object
marital            object
education          object
default            object
housing            object
loan               object
contact            object
month              object
day_of_week        object
duration            int64
campaign            int64
pdays               int64
previous            int64
poutcome           object
emp.var.rate      float64
cons.price.idx    float64
cons.conf.idx     float64
euribor3m         float64
nr.employed       float64
y                  object
dtype: object

In [5]:
df.describe()

age      duration  ...     euribor3m   nr.employed
count  41188.00000  41188.000000  ...  41188.000000  41188.000000
mean      40.02406    258.285010  ...      3.621291   5167.035911
std       10.42125    259.279249  ...      1.734447     72.251528
min       17.00000      0.000000  ...      0.634000   4963.600000
25%       32.00000    102.000000  ...      1.344000   5099.100000
50%       38.00000    180.000000  ...      4.857000   5191.000000
75%       47.00000    319.000000  ...      4.961000   5228.100000
max       98.00000   4918.000000  ...      5.045000   5228.100000

[8 rows x 10 columns]

In [6]:
# Check for missing values
df.isnull().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64

In [7]:
# Categorical Columns (indices) - [1:9]
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
# le_columns = X[:, [1, 2, 3, 4, 5, 6, 7, 8, 9, 14]]

# for i in X[:]:
#   if i in le_columns:
#     X[:, i] = le.fit_transform(X[:, i])

X[:, 1] = le.fit_transform(X[:, 1])
X[:, 2] = le.fit_transform(X[:, 2])
X[:, 3] = le.fit_transform(X[:, 3])
X[:, 4] = le.fit_transform(X[:, 4])
X[:, 5] = le.fit_transform(X[:, 5])
X[:, 6] = le.fit_transform(X[:, 6])
X[:, 7] = le.fit_transform(X[:, 7])
X[:, 8] = le.fit_transform(X[:, 8])
X[:, 9] = le.fit_transform(X[:, 9])
X[:, 14] = le.fit_transform(X[:, 14])
y = le.fit_transform(y)

print(X[:, 1:14])
print(y)

[[3 1 0 ... 1 999 0]
 [7 1 3 ... 1 999 0]
 [7 1 3 ... 1 999 0]
 ...
 [5 1 6 ... 2 999 0]
 [9 1 5 ... 1 999 0]
 [5 1 5 ... 3 999 1]]
[0 0 0 ... 0 1 0]


In [8]:
# Split dataframe (df) into training and test sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, 
                                                    random_state=0)

In [9]:
# Feature Scale numerical columns
from sklearn.preprocessing import StandardScaler, MinMaxScaler

# Apply Standardization
numerical_columns = [0, 10, 11, 12, 13, 15, 16, 17, 18, 19]
sc = StandardScaler()
X_train[:, numerical_columns] = sc.fit_transform(X_train[:, numerical_columns])
X_test[:, numerical_columns] = sc.transform(X_test[:, numerical_columns])

## Define XGB model

In [10]:
import xgboost as xgb
from xgboost import XGBClassifier

# Accuracy Scoring
clf = XGBClassifier(max_depth=6, n_estimators=99, learning_rate=0.1, 
                    colsample_bytree=0.5, gamma=0, min_child_weight=1, 
                    reg_alpha=0.1, subsample=1, random_state=0)

# ROC_AUC Scoring
# clf = XGBClassifier(max_depth=6, n_estimators=104, learning_rate=0.09, 
#                     gamma=0.02, colsample_bytree=0.7, min_child_weight=5,
#                     reg_alpha=1.2, subsample=1, random_state=7)

# Train Model
clf.fit(X_train, y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.5, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=None, n_estimators=99, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0.1, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

## Evaluate model

In [11]:
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.metrics import accuracy_score, roc_auc_score

# Predict test set results
# If y_pred > 0.5, client has subscribed to a term deposit
y_pred = clf.predict(X_test)
y_pred = (y_pred > 0.5)
print('y_pred: y_test:')
print(np.concatenate((y_pred.reshape(len(y_pred), 1), y_test.reshape(len(y_test), 1)), 1 ))
print('')

# Create the confusion matrix
# TN = cm[0][0]  # True Negative
# FP = cm[0][1]  # False Positive
# FN = cm[1][0]  # False Negative
# TP = cm[1][1]  # True Positive
print('Confusion Matrix: ')
cm = confusion_matrix(y_test, y_pred)
print(cm)
print('')
print("Test Accuracy: %.3f%% " % (accuracy_score(y_test, y_pred) * 100))
# print("ROC_AUC: %.3f%% " % (roc_auc_score(y_test, y_pred) * 100))
print('')

# accuracy = (TP + TN)/(TN + FP + FN + TP)
# precision = TP/(TP + FP)
# recall = TP/(TP + FN)
# f1_score = (2 * precision * recall)/(precision + recall)
# support = Number of occurrences of each class in y_pred
print('Classification Report: ')
print(classification_report(y_test, y_pred))


y_pred: y_test:
[[0 1]
 [0 0]
 [0 0]
 ...
 [1 1]
 [0 0]
 [0 0]]

Confusion Matrix: 
[[7081  238]
 [ 401  518]]

Test Accuracy: 92.243% 

Classification Report: 
              precision    recall  f1-score   support

           0       0.95      0.97      0.96      7319
           1       0.69      0.56      0.62       919

    accuracy                           0.92      8238
   macro avg       0.82      0.77      0.79      8238
weighted avg       0.92      0.92      0.92      8238



In [12]:
# Apply k-Fold Cross Validation on the training set
from sklearn.model_selection import cross_val_score, RepeatedStratifiedKFold

cv = RepeatedStratifiedKFold(n_splits = 10, n_repeats = 3, random_state = 0)
scores = cross_val_score(clf, X_train, y_train, scoring='accuracy', cv=cv, n_jobs=-1)
# scores = cross_val_score(clf, X_train, y_train, scoring='roc_auc', cv=cv, n_jobs=-1)

# Force scores to be positive
scores = abs(scores)
print("Mean Accuracy: %.3f%%" % (np.mean(scores) * 100))
print("Standard Deviation: %.3f%%" % np.std(scores))
print("Min Accuracy: %.3f%%" % (np.min(scores) * 100))
print("Max Accuracy: %.3f%%" % (np.max(scores) * 100))

Mean Accuracy: 91.689%
Standard Deviation: 0.004%
Min Accuracy: 90.986%
Max Accuracy: 92.504%


<table border = "1">
  <thead>
    <th colspan = "3">Cross-Validation Results</th>
  </thead>
  <tbody>
    <tr>
      <td><b></b></td>
      <td><b>Accuracy</b></td>
      <td><b>ROC_AUC</b></td>
    </tr>
    <tr>
      <td><b>Mean</b></td>
      <td>91.689%</td>
      <td>94.950%</td>
    </tr>
    <tr>
      <td><b>Std.</b></td>
      <td>0.004%</td>
      <td>0.004%</td>
    </tr>
    <tr>
      <td><b>Min</b></td>
      <td>90.986%</td>
      <td>94.132%</td>
    </tr>
    <tr>
      <td><b>Max</b></td>
      <td>92.504%</td>
      <td>95.564%</td>
    </tr>
  </tbody>
</table>

In [13]:
# Apply Grid Search for parameter tuning
from sklearn.model_selection import GridSearchCV

# param_dict = {'max_depth': [1, 2, 5, 10, 20], 
#               'n_estimators': [50, 100, 150, 200],
#               'learning_rate': [0.05, 0.1, 0.15],
#               'colsample_bytree': [0.3, 0.5, 0.7],
#               'gamma': [0, 0.05],
#               'min_child_weight': [2, 5, 8, 10],
#               'reg_alpha':[0.001, 0.01, 0.1, 1, 10],
#               'subsample': [0.6, 0.8, 1.0],
#               'random_state': [0, 1, 5, 10, 20, 40]}

# Accuracy scoring
param_dict = {'max_depth': [6],
              'n_estimators': [99],
              'learning_rate': [0.1],
              'colsample_bytree': [0.5],
              'gamma': [0],
              'min_child_weight': [1],
              'reg_alpha':[0.1],
              'subsample': [1.0],
              'random_state': [0, 1]}

# ROC_AUC scoring
# param_dict = {'max_depth': [6],
#               'n_estimators': [104],
#               'learning_rate': [0.09],
#               'colsample_bytree': [0.7],
#               'gamma': [0.02],
#               'min_child_weight': [5],
#               'reg_alpha':[1.2],
#               'subsample': [1],
#               'random_state': [6, 7]}

grid = GridSearchCV(clf, param_dict, scoring='accuracy', cv=cv, n_jobs=-1)
# grid = GridSearchCV(clf, param_dict, scoring='roc_auc', cv=cv, n_jobs=-1)
results = grid.fit(X_train, y_train)

# Summarize results
best_accuracy = grid.best_score_
best_parameters = grid.best_params_
print("Best Accuracy: %.3f%%" % (best_accuracy * 100))
print("Best Parameters: ", best_parameters)


Best Accuracy: 91.689%
Best Parameters:  {'colsample_bytree': 0.5, 'gamma': 0, 'learning_rate': 0.1, 'max_depth': 6, 'min_child_weight': 1, 'n_estimators': 99, 'random_state': 0, 'reg_alpha': 0.1, 'subsample': 1.0}


<table border = "1">
  <thead>
    <th colspan = "3">Grid Search Results</th>
  </thead>
  <tbody>
    <tr>
      <td><b></b></td>
      <td><b>Accuracy</b></td>
      <td><b>ROC_AUC</b></td>
    </tr>
    <tr>
      <td><b>Best Score</b></td>
      <td>91.689%</td>
      <td>94.950%</td>
    </tr>
    <tr>
      <td colspan="3"><b>Best Parameters:</b></td>
    </tr>
    <tr>
      <td><b>max_depth</b></td>
      <td>6</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>n_estimators</b></td>
      <td>99</td>
      <td>104</td>
    </tr>
    <tr>
      <td><b>learning_rate</b></td>
      <td>0.1</td>
      <td>0.09</td>
    </tr>
    <tr>
      <td><b>colsample_bytree</b></td>
      <td>0.5</td>
      <td>0.7</td>
    </tr>
    <tr>
      <td><b>gamma</b></td>
      <td>0</td>
      <td>0.02</td>
    </tr>
    <tr>
      <td><b>min_child_weight</b></td>
      <td>1</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>reg_alpha</b></td>
      <td>0.1</td>
      <td>1.2</td>
    </tr>
    <tr>
      <td><b>subsample</b></td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>random_state</b></td>
      <td>0</td>
      <td>7</td>
    </tr>
  </tbody>
</table>